File: app/models/payment.py
"""Database model for payments."""

from datetime import datetime
from decimal import Decimal
from typing import Optional
from enum import Enum

from pydantic import BaseModel, Field, ConfigDict


class PaymentStatus(str, Enum):
    """Payment status enumeration."""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    REFUNDED = "refunded"


class PaymentMethod(str, Enum):
    """Payment method enumeration."""
    CARD = "card"
    BANK_TRANSFER = "bank_transfer"
    WALLET = "wallet"
    UPI = "upi"
    CRYPTO = "crypto"


class Payment(BaseModel):
    """Represents a payment transaction."""

    model_config = ConfigDict(use_enum_values=True)

    id: str = Field(..., description="Payment identifier")
    amount: Decimal = Field(..., ge=0, description="Payment amount")
    currency: str = Field(..., description="Currency code (ISO 4217)")
    status: PaymentStatus = Field(default=PaymentStatus.PENDING, description="Payment status")
    method: PaymentMethod = Field(..., description="Payment method")
    customer_id: str = Field(..., description="Customer identifier")
    provider: str = Field(..., description="Payment provider")
    provider_transaction_id: Optional[str] = Field(None, description="Provider transaction ID")
    description: Optional[str] = Field(None, description="Payment description")
    metadata: Optional[dict] = Field(None, description="Additional metadata")
    created_at: Optional[datetime] = Field(None, description="Creation timestamp")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")

File: frontend/src/types/payment.ts
export interface Payment {
  id: string;
  customer_id: string;
  amount: number;
  currency: string;
  method: PaymentMethod;
  status: PaymentStatus;
  description?: string;
  metadata?: Record<string, any>;
  provider: PaymentProvider;
  provider_payment_id?: string;
  idempotency_key: string;
  created_at: string;
  updated_at: string;
  completed_at?: string;
  failed_at?: string;
  error_message?: string;
}

export enum PaymentStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
  REFUNDED = 'refunded'
}

export enum PaymentMethod {
  CARD = 'card',
  BANK_TRANSFER = 'bank_transfer',
  WALLET = 'wallet',
  UPI = 'upi'
}

export enum PaymentProvider {
  STRIPE = 'stripe',
  RAZORPAY = 'razorpay'
}

export interface CreatePaymentRequest {
  amount: number;
  currency: string;
  method: PaymentMethod;
  customer_id: string;
  description?: string;
  metadata?: Record<string, any>;
}

export interface PaymentResponse {
  success: boolean;
  payment?: Payment;
  error?: string;
  payment_url?: string;
}

export interface RefundRequest {
  payment_id: string;
  amount?: number;
  reason: string;
}

export interface PaymentMetrics {
  total_transactions: number;
  total_revenue: number;
  success_rate: number;
  average_amount: number;
  transactions_by_method: Record<PaymentMethod, number>;
  transactions_by_status: Record<PaymentStatus, number>;
  revenue_by_currency: Record<string, number>;
}

export interface Customer {
  id: string;
  email?: string;
  phone?: string;
  name?: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
} 